src/main/java/io/aiven/kafka/connect/http/RecordSender.java

```diff
@@ -39,88 +34,40 @@
+        for (final SinkRecord record : records) {
+            // TODO proper batching
+            final RecordBatch batch = new RecordBatch(List.of(record));
+            sendBatchWithRetries(batch);
+        }
     }
 
-    void stop() {
-        executorService.shutdownNow();
-    }
+    private void sendBatchWithRetries(final RecordBatch batch) throws InterruptedException {
+        // TODO proper batching
+        final String body = (String) batch.records().get(0).value();
 
-    private final class BackgroundSender implements Runnable {
-        @Override
-        public void run() {
+        int remainRetries = RecordSender.this.maxRetries;
+        while (true) {
             try {
-                while (true) {
-                    final SinkRecord record = outstandingRecords.poll(1, TimeUnit.SECONDS);
-                    if (record == null) {
-                        continue;
-                    }
-
-                    // TODO proper batching
-                    final RecordBatch batch = new RecordBatch(List.of(record));
-                    if (!sendBatchWithRetries(batch)) {
-                        return;
-                    }
-
-                    lastSentOffsets.putAll(batch.latestOffsets());
-                }
-            } catch (final InterruptedException e) {
-                log.info("Interrupted", e);
-            }
-        }
-
-        private boolean sendBatchWithRetries(final RecordBatch batch) throws InterruptedException {
-            log.debug("Sending batch {}", batch);
-
-            // TODO proper batching
-            final String body = (String) batch.records().get(0).value();
-
-            int remainRetries = RecordSender.this.maxRetries;
-            while (true) {
-                try {
-                    httpSender.send(body);
-                    return true;
-                } catch (final IOException e) {
-                    if (remainRetries > 0) {
-                        log.info("Sending batch {} failed, will retry in {} ms ({} retries remain)",
-                            batch, RecordSender.this.retryBackoffMs, remainRetries, e);
-                        remainRetries -= 1;
-                        Thread.sleep(RecordSender.this.retryBackoffMs);
-                    } else {
-                        log.error("Sending batch {} failed and no retries remain, stopping",
-                            batch);
-                        sendException = e;
-                        return false;
-                    }
+                httpSender.send(body);
+                return;
+            } catch (final IOException e) {
+                if (remainRetries > 0) {
+                    log.info("Sending batch failed, will retry in {} ms ({} retries remain)",
+                        RecordSender.this.retryBackoffMs, remainRetries, e);
+                    remainRetries -= 1;
+                    Thread.sleep(RecordSender.this.retryBackoffMs);
```

Comment:
Does it make sense to use additive increase, multiplicative decrease (AIMD) approach here?

Reply:
Some better backoff algorithm is needed for sure (like a port of [this](https://github.com/aiven/aiven-core/blob/dd7a1ab1454dc0f12a5800c677955575a717da57/aiven/shared/utils.py#L532)). Will be added at some point :)